### PR TITLE
 Fix issue with GetSimpleNonTypeMembers on records

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/49302 (incorrectly resolving to object equality operator)
Closes https://github.com/dotnet/roslyn/issues/49263

Filled https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1273244 for ask-mode approval.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:279]: Does this call `GetSimpleNonTypeMembers()`? Consider calling that method directly.

Same comment for test below.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:278]: Could use `lib_comp.AsReference(useCompilationReference)`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:279]: Both tests repro the issue if the fix isn't present. The one with multiple compilations is from the reported issue, the one with single compilation is more direct.

**Review by cston — APPROVED** [review]

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. Thanks

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:249]: I can already see this behavior on the current master with SharpLab.

image

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:232]: This test is an exact duplicate of `ToString_SelfReferentialBase` test in the same file.

**Comment by Youssef1313** [comment]
The last commit here was **after merging** https://github.com/dotnet/roslyn/pull/50848. The 32-bit shouldn't have failed.

@sharwell Could you take a look?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs]: Removed the duplicate test. Thanks

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1323]: >declaration.Kind == DeclarationKind.Record [](start = 92, length = 42)

Is it possible to make this check more precise by actually checking if the ```name``` among those names that could be added for records?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:279]: > Does this call GetSimpleNonTypeMembers()? Consider calling that method directly.

Agree with this suggestion.

---
In reply to: [567031521](https://github.com/dotnet/roslyn/pull/50900#discussion_r567031521) [](ancestors = 567031521)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1323]: It's possible, but it would add complexity due to properties, their accessors and their backing fields. I don't think it's worthwhile.

---
In reply to: [567952525](https://github.com/dotnet/roslyn/pull/50900#discussion_r567952525) [](ancestors = 567952525)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:283]: Is this type used? Also, I would simply include this check into the previous test.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:283]: The reason for separate tests  is that we want to make sure to hit the code before members have been pulled on.
I'll remove type A

**Comment by jcouv** [comment]
QB approved the bug/fix. Marking as approved.

**Comment by sharwell** [comment]
> The last commit here was after merging #50848. The 32-bit integration test job shouldn't have failed.
>
> @sharwell Could you take a look?

The build job checked out commit 38e27262e087ac1ede2117cc32c51c5b033fe6da (seen in the Shallow Checkout log) which does not contain my change.

https://github.com/dotnet/roslyn/blob/38e27262e087ac1ede2117cc32c51c5b033fe6da/eng/build.ps1#L422-L425

**Pull request merged**
